File: lan_folder_sync/tests/dry_run_test.py
import json
import os
import socket
import subprocess

from framing import recv_msg, send_msg

HOST, PORT = "127.0.0.1", 8765
ROOT_DIR = "sandbox/source"
PROBE = f"_dryrun_probe_{os.getpid()}.txt"  # unique so it can't pre-exist on peer


def remote_manifest():
    """Ask the live server what files its received"""
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((HOST, PORT))
    send_msg(s, json.dumps({"op": "MANIFEST"}).encode("utf-8"))
    reply = recv_msg(s)
    s.close()
    assert reply is not None, "server closed without sending a manifest"
    return json.loads(reply.decode("utf-8"))["files"]


# 1. Plant a file locally that the peer definitely does not have
os.makedirs(ROOT_DIR, exist_ok=True)
probe_path = os.path.join(ROOT_DIR, PROBE)
with open(probe_path, "w") as f:
    f.write("dry-run probe - must never reach the peer\n")

try:
    # 2. Sanity: the peer starts without it
    assert PROBE not in remote_manifest(), f"peer already has {PROBE}?!"

    # 3. Run the real push, in dry-run, and capture what it prints
    result = subprocess.run(
        ["python3", "sync_push.py", HOST, ROOT_DIR, "--dry-run"],
        capture_output=True,
        text=True,
        check=False,
    )
    print(result.stdout, end="")
    assert result.returncode == 0, f"push crashed:\n{result.stderr}"

    # 4. It should plan to send the probe...
    assert "DRY RUN" in result.stdout, "no DRY RUN banner in output"
    assert PROBE in result.stdout, f"{PROBE} not listed in the plan"

    # 5. ...but the peer must still not have it. The plan was talk, not action
    assert PROBE not in remote_manifest(), (
        f"{PROBE} reached the peer - dry-run transferred a file!"
    )

    print(f"\nOK dry-run planned PUT {PROBE} but transferred nothing")
finally:
    os.remove(probe_path)

File: lan_folder_sync_c/tests/dry_run_test.c
/* --dry-run prints the plan and transfers nothing. Mirrors dry_run_test.py. */
#include "harness.h"

#define ROOT_DIR "sandbox/source"

int main(void)
{
    char probe[64], probe_path[256];
    snprintf(probe, sizeof(probe), "_dryrun_probe_%d.txt", (int)getpid());
    snprintf(probe_path, sizeof(probe_path), "%s/%s", ROOT_DIR, probe);

    section("Dry-run");

    /* 1. Plant a file locally that the peer definitely does not have */
    write_text_file(probe_path, "dry-run probe - must never reach the peer\n");

    /* 2. Sanity: the peer starts without it */
    manifest before;
    remote_manifest(&before);
    CHECK(!manifest_has(&before, probe), "peer already has %s?!", probe);
    manifest_free(&before);

    /* 3. Run the real push, in dry-run, and capture what it prints */
    char cmd[512];
    snprintf(cmd, sizeof(cmd), "./bin/sync_push %s %s --dry-run 2>&1", TEST_HOST,
             ROOT_DIR);
    int status = 0;
    char *out = run_capture(cmd, &status);
    info(out);
    CHECK(status == 0, "push crashed (exit %d)", status);

    /* 4. It should plan to send the probe... */
    CHECK(strstr(out, "DRY RUN") != NULL, "no DRY RUN banner in output");
    CHECK(strstr(out, probe) != NULL, "%s not listed in the plan", probe);

    /* 5. ...but the peer must still not have it. The plan was talk, not action */
    manifest after;
    remote_manifest(&after);
    CHECK(!manifest_has(&after, probe),
          "%s reached the peer - dry-run transferred a file!", probe);
    manifest_free(&after);

    ok("dry-run planned PUT %s but transferred nothing", probe);
    free(out);
    unlink(probe_path);

    done("dry-run previews without touching the peer");
    return 0;
}
